File: forms/cardiology-request/front-end-with-svelte/src/lib/engine/safety-rules.ts
import type { CardiologyRequest, SafetyBand, FiredRule } from './types';
import { hasRedFlag, hasTypicalAngina } from './utils';

/**
 * Axis B — safety / red-flag.
 *
 * Escalation ladder (ok → caution → red-flag):
 * - red-flag: any acute red flag (suspected acute coronary syndrome, exertional
 *   syncope, new-onset heart failure), an elevated troponin, or typical-angina
 *   chest pain. These drive the safety axis and auto-escalate the triage tier.
 * - caution: softer acuity signals (elevated BNP, NYHA III–IV breathlessness,
 *   syncope, palpitations with no red flag).
 * - ok: no safety signal.
 *
 * The least-alarming band is chosen only when no rule fires.
 */
export function gradeSafety(r: CardiologyRequest): {
	safetyBand: SafetyBand;
	firedRules: FiredRule[];
} {
	const firedRules: FiredRule[] = [];

	// ─── red-flag ───
	if (r.suspectedAcs) {
		firedRules.push({
			ruleId: 'R-SAFETY-ACS-01',
			axis: 'safety',
			category: 'suspected-acs',
			description:
				'Suspected acute coronary syndrome; emergency pathway warranted, not a routine cardiology referral.'
		});
	}
	if (r.exertionalSyncope) {
		firedRules.push({
			ruleId: 'R-SAFETY-EXERTIONAL-SYNCOPE-01',
			axis: 'safety',
			category: 'exertional-syncope',
			description:
				'Exertional syncope; possible structural or arrhythmic cause requiring urgent assessment.'
		});
	}
	if (r.newOnsetHeartFailure) {
		firedRules.push({
			ruleId: 'R-SAFETY-NEW-HF-01',
			axis: 'safety',
			category: 'new-onset-heart-failure',
			description: 'New-onset heart failure; warrants urgent assessment per NICE NG106.'
		});
	}
	if (r.troponinStatus === 'elevated') {
		firedRules.push({
			ruleId: 'R-SAFETY-TROPONIN-01',
			axis: 'safety',
			category: 'elevated-troponin',
			description: 'Elevated troponin recorded; treat as an acute coronary presentation.'
		});
	}
	if (hasTypicalAngina(r)) {
		firedRules.push({
			ruleId: 'R-SAFETY-TYPICAL-ANGINA-01',
			axis: 'safety',
			category: 'red-flag-chest-pain',
			description: 'Typical-angina chest pain; vet for urgent rather than routine review.'
		});
	}

	if (hasRedFlag(r) || r.troponinStatus === 'elevated' || hasTypicalAngina(r)) {
		return { safetyBand: 'red-flag', firedRules };
	}

	// ─── caution ───
	const cautionSignals =
		r.bnpStatus === 'elevated' ||
		r.nyhaClass === 'iii' ||
		r.nyhaClass === 'iv' ||
		r.symptomSyncope ||
		r.symptomPalpitations;

	if (cautionSignals) {
		firedRules.push({
			ruleId: 'R-SAFETY-CAUTION-01',
			axis: 'safety',
			category: 'acuity-signal',
			description:
				'Softer acuity signal present (elevated BNP, NYHA III–IV, syncope, or palpitations); manage with caution.'
		});
		return { safetyBand: 'caution', firedRules };
	}

	// ─── ok: least-alarming band, no rule fired ───
	firedRules.push({
		ruleId: 'R-SAFETY-OK-01',
		axis: 'safety',
		category: 'no-red-flag',
		description: 'No red flag or acuity signal; safety band OK.'
	});
	return { safetyBand: 'ok', firedRules };
}

File: forms/cardiology-request/back-end-with-loco/src/engine/safety_rules.rs
//! Axis B — safety / red-flag.
//!
//! Ported from `front-end-with-svelte/src/lib/engine/safety-rules.ts`.

use super::types::{CardiologyRequest, FiredRule};
use super::utils::{has_red_flag, has_typical_angina};

/// The result of grading axis B.
pub struct Safety {
    /// The safety band (`ok` / `caution` / `red-flag`).
    pub safety_band: String,
    /// The audit-trail rules that fired.
    pub fired_rules: Vec<FiredRule>,
}

/// Grade axis B — safety / red-flag, on the escalation ladder
/// `ok → caution → red-flag`. The least-alarming band is chosen only when no
/// rule fires.
#[must_use]
pub fn grade_safety(r: &CardiologyRequest) -> Safety {
    let mut fired_rules: Vec<FiredRule> = Vec::new();

    // ─── red-flag ───
    if r.suspected_acs {
        fired_rules.push(FiredRule::new(
            "R-SAFETY-ACS-01",
            "safety",
            "suspected-acs",
            "Suspected acute coronary syndrome; emergency pathway warranted, not a routine cardiology referral.",
        ));
    }
    if r.exertional_syncope {
        fired_rules.push(FiredRule::new(
            "R-SAFETY-EXERTIONAL-SYNCOPE-01",
            "safety",
            "exertional-syncope",
            "Exertional syncope; possible structural or arrhythmic cause requiring urgent assessment.",
        ));
    }
    if r.new_onset_heart_failure {
        fired_rules.push(FiredRule::new(
            "R-SAFETY-NEW-HF-01",
            "safety",
            "new-onset-heart-failure",
            "New-onset heart failure; warrants urgent assessment per NICE NG106.",
        ));
    }
    if r.troponin_status == "elevated" {
        fired_rules.push(FiredRule::new(
            "R-SAFETY-TROPONIN-01",
            "safety",
            "elevated-troponin",
            "Elevated troponin recorded; treat as an acute coronary presentation.",
        ));
    }
    if has_typical_angina(r) {
        fired_rules.push(FiredRule::new(
            "R-SAFETY-TYPICAL-ANGINA-01",
            "safety",
            "red-flag-chest-pain",
            "Typical-angina chest pain; vet for urgent rather than routine review.",
        ));
    }

    if has_red_flag(r) || r.troponin_status == "elevated" || has_typical_angina(r) {
        return Safety {
            safety_band: "red-flag".to_string(),
            fired_rules,
        };
    }

    // ─── caution ───
    let caution_signals = r.bnp_status == "elevated"
        || r.nyha_class == "iii"
        || r.nyha_class == "iv"
        || r.symptom_syncope
        || r.symptom_palpitations;

    if caution_signals {
        fired_rules.push(FiredRule::new(
            "R-SAFETY-CAUTION-01",
            "safety",
            "acuity-signal",
            "Softer acuity signal present (elevated BNP, NYHA III–IV, syncope, or palpitations); manage with caution.",
        ));
        return Safety {
            safety_band: "caution".to_string(),
            fired_rules,
        };
    }

    // ─── ok: least-alarming band, no rule fired ───
    fired_rules.push(FiredRule::new(
        "R-SAFETY-OK-01",
        "safety",
        "no-red-flag",
        "No red flag or acuity signal; safety band OK.",
    ));
    Safety {
        safety_band: "ok".to_string(),
        fired_rules,
    }
}
